### PR TITLE
router: add upstream access log option on upstream stream

### PR DESCRIPTION
**Pull request by ohadvano**
Commit Message:
Additional Description: Follow up to https://github.com/envoyproxy/envoy/pull/26155, enabling a similar feature for upstream access logs - record an access log on upstream connection established, if requested.
Risk Level: Low
Testing: Unit tests, integration tests
Docs Changes: router API
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

File: source/common/router/upstream_request.cc
Comment (lizan) [inline source/common/router/upstream_request.cc:299]: Is 2xx the only case we want to mark it InProgress? How about other codes with a body? I'm not sure whether we want to depend on the status code.

**Review by lizan — COMMENTED** [review]
Generally LGTM, one question

File: source/common/router/upstream_request.cc
Comment (ohadvano) [reply source/common/router/upstream_request.cc:299]: Setting InProgress is mainly to be able to distinguish access logs in mid streams (using periodic access logs). Therefore, I think that setting InProgress in cases the stream will closed after response status code might confuse a user.

There would be a follow up PR to add period upstream access logs (similar to HCM periodic upstream access logs), then consider the following case:
The HTTP response headers had returned with 400 status code, the stream state is set to InProgress, and then an access log is recorded due to timer triggered. The output will include InProgress state, although the stream is terminating.

WDYT?

Comment (ohadvano) [reply source/common/router/upstream_request.cc]: After an offline discussion, we can to an agreement that in some cases a very large response body or a network delay may cause a time gap between the headers response and finish of the body encoding, regardless of the HTTP status code. Therefore, removed the status code check and keeping only "!end_stream"

File: api/envoy/extensions/filters/http/router/v3/router.proto
Comment (alyssawilk) [inline api/envoy/extensions/filters/http/router/v3/router.proto:30]: this could be misread as 2 additional logs, one on new stream and one after connection
maybe rephrase "will be recorded when an upstream stream is associated to an http request"?

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:22]: TCP connection or HTTP stream I think

File: changelogs/current.yaml
Comment (alyssawilk) [inline changelogs/current.yaml:358]: nit: upstream stream is created?

for HTTP/1 these are one and the same but for H/2 we may have a preestablished connection, and so I want to be clear :-)

File: source/common/router/upstream_request.cc
Comment (alyssawilk) [inline source/common/router/upstream_request.cc:208]: I'm confused why we need this.
One can already infer most of this from various timing characteristics and I think having a separate round of state we need to keep track of is problematic
I see this isn't technically a part of this PR but landed in https://github.com/envoyproxy/envoy/pull/26256 but that somehow was merged without senior maintainer sign-off and I think we may need to revisit it and wait on this one until that's sorted out.

File: source/common/router/upstream_request.cc
Comment (alyssawilk) [inline source/common/router/upstream_request.cc:226]: isn't upstream_headers.get() simply null if the pointer is not set

**Review by alyssawilk — COMMENTED** [review]
drive by comments :-)

File: source/common/router/upstream_request.cc
Comment (ohadvano) [reply source/common/router/upstream_request.cc:208]: The follow PR: https://github.com/envoyproxy/envoy/pull/26203 change the behavior of %DURATION% so it's now returning a value at any time queried during a stream. When periodic access logs were originally added, Matt had a concern about not having a way to distinguish between an end of stream access log and mid-stream access log. Therefore, instead of %DURATION% (that was null for mid-stream access logs), %STEAM_STATE% was introduced, which returns either Started, InProgress or Ended. The streamState() is currently only being used for this purpose. Please let me know if this makes more sense now

File: source/common/router/upstream_request.cc
Comment (ohadvano) [reply source/common/router/upstream_request.cc]: done

File: changelogs/current.yaml
Comment (ohadvano) [reply changelogs/current.yaml]: done

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst]: done

File: source/common/router/upstream_request.cc
Comment (alyssawilk) [inline source/common/router/upstream_request.cc:296]: docs in your prior PR indicate 
The value will be 'InProgress' in case an upstream connection was made,

This is being set on upstream headers being received which can be well after the stream is established. We'd also have to handle the case where we have an upstream stream, get an error response, and disassociate the stream for a pending retry  (ad may not get another immediae upstream connection).

I think we should scrap the state in stream info, and infer the successful upstream connection from the presence of an upstream host, making sure we set the state properly in access logging based on already existing stream info properties.

**Review by alyssawilk — CHANGES_REQUESTED** [review]

File: source/common/router/upstream_request.cc
Comment (ohadvano) [reply source/common/router/upstream_request.cc]: @alyssawilk, I have deleted usage of StreamState in this PR. After this one is merged I'll create another one to adjust the code per what we discussed in the other PR. Can you please have another look?

Comment (alyssawilk) [reply source/common/router/upstream_request.cc]: As I said above I think we need to work out the logging APIs before we land this. The release is cut next week and if we don't fix the logging before then we will have to either 1) roll back the other PR or 2) support the non-extensible APIs perpetually.  I would lean towards 1 so I do not want any dependant PRs landing.
/wait

Comment (ohadvano) [reply source/common/router/upstream_request.cc]: @alyssawilk can we merge this one, now that the revert PR is merged?

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:20]: should we update this text to remove references to TCP which is now no longer implemented?

File: api/envoy/extensions/filters/http/router/v3/router.proto
Comment (alyssawilk) [inline api/envoy/extensions/filters/http/router/v3/router.proto:35]: Should this be a boolean or an optional AccessLog message?  I could imagine folks wanting to send their connection established logs to a different endpoint than standard upstream access logs though I'll defer to @lizan as API reviewer for that.

mild preference for on_upstream_stream rather than new_request, as request/response is minimally used in Envoy

File: test/common/router/router_upstream_log_test.cc
Comment (alyssawilk) [inline test/common/router/router_upstream_log_test.cc:419]: I'd definitely like an e2e test both of the standard case and a second log for retried request.  waitForAccessLog should work as long as you make sure to use the optional fields to wait for the right log entry.

**Review by alyssawilk — COMMENTED** [review]
yeah this can land ahead of fixing up the access log state to include which logger it came from.  Here's my thoughts for this pass.  I'm out tomorrow so will pick up the review Monday but I can pick this up and the roll-forward of one of the others in parallel if you'd like.

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:20]: Not sure what you mean, what is no longer implemented in TCP?

File: api/envoy/extensions/filters/http/router/v3/router.proto
Comment (ohadvano) [reply api/envoy/extensions/filters/http/router/v3/router.proto:35]: > Should this be a boolean or an optional AccessLog message? I could imagine folks wanting to send their connection established logs to a different endpoint than standard upstream access logs though I'll defer to @lizan as API reviewer for that.

Is it not possible for the user to use the existing AccessLog field with another sink?

> mild preference for on_upstream_stream rather than new_request, as request/response is minimally used in Envoy

Makes sense, I'll update it

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [reply docs/root/intro/arch_overview/observability/access_logging.rst:20]: ah apologies, for some reason I thought the TCP log stream info were done in the same PR as the tcp log flushing.  the suggestion was in error and docs are fine as is.

File: test/common/router/router_upstream_log_test.cc
Comment (ohadvano) [reply test/common/router/router_upstream_log_test.cc:419]: I'm not sure how to cause retry for the router in integration tests. Is there even currently an integration test that checks that the standard upstream log is flushed twice for retry?
Anyways, I did add a check to integration tests, that if the bool config is enabled, then there are two upstream access log made.

Comment (ohadvano) [reply test/common/router/router_upstream_log_test.cc:419]: Never mind, figured it out. I renamed ``upstream_filter_state_integration_test.cc`` to ``upstream_access_log_integration_test.cc`` since it only had a single test, that was tightly coupled with upstream access logs, and added a test that checks an upstream access log with retry (total of 4 logs - start, end repeated)

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:46]: when a new HTTP request is received -> when a new upstream stream is associated with the downstream stream

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:60]: upstream request -> upstream connection

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (alyssawilk) [inline docs/root/intro/arch_overview/observability/access_logging.rst:67]: If you have both periodic and "flush on new request" enabled, will we do 2 logs on downstream connection or only one?

**Review by alyssawilk — COMMENTED** [review]
cool LGTM modulo a few docs nits.   @zuercher could you do second pass?  Also @lizan this is still waiting on API review.

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:46]: done

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst]: done

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst:68]: I wrote above the "Start of session access logs" that: "Reporting access logs right after upstream connection establishment or new incoming HTTP request does not depend on periodic reporting, and the other way around." once, otherwise I'd have to repeat this for both sections. Please let me know if further elaboration is required.

Comment (zuercher) [reply docs/root/intro/arch_overview/observability/access_logging.rst:68]: I don't think it needs to be repeated in each section, but it wasn't immediately clear to me that enabling both features would cause two log messages. (I think it would be nice if we detected both features were on and only logged once at the start, but I haven't looked at the periodic code closely enough to know how difficult that would be.)

**Review by zuercher — COMMENTED** [review]
This looks good, but I do have one related concern. Since logs historically were only generated at the end of a request of connection I worry that there may be formatters that trigger undefined behavior when invoked before the end of the request. I wonder if there's a way we can add a test that exercises them in the context of this feature (so that our various thread/memory sanitizer builds can do validate them).

**Review by lizan — COMMENTED** [review]
/lgtm api

**Comment by ohadvano** [comment]
> This looks good, but I do have one related concern. Since logs historically were only generated at the end of a request of connection I worry that there may be formatters that trigger undefined behavior when invoked before the end of the request. I wonder if there's a way we can add a test that exercises them in the context of this feature (so that our various thread/memory sanitizer builds can do validate them).

@zuercher thanks for your review, however I'm not sure I follow through. As far as I can tell, the formatter action is decoupled from the access log operation, and sometimes being used orthogonally regardless of access log flow. The formatter has a lot of substitution string options, and they are being tested in a dedicated test scope. Maybe if you can show an example of what kind of test could help here, I'll be able to understand more of this. Might worth mentioning that this implementation is similar in its behavior to how start of stream/request access logs were added to TcpProxy and HCM.

**Comment by alyssawilk** [comment]
@zuercher FWIW the new integration does test this for the newly added log trigger so I think we've at least got some coverage for your *san concerns.

**Comment by zuercher** [comment]
So thinking about it more, all the log formatting is done using the same mechanisms as for HTTP header manipulation, so they likely don't have issues executing mid-request.

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (zuercher) [reply docs/root/intro/arch_overview/observability/access_logging.rst:68]: I see. I added a related comment that I think would make the behavior clear to me.

File: docs/root/intro/arch_overview/observability/access_logging.rst
Comment (zuercher) [reply docs/root/intro/arch_overview/observability/access_logging.rst:61]: also nit: "new connection is received"

Comment (zuercher) [inline docs/root/intro/arch_overview/observability/access_logging.rst:61]:
```suggestion
Note: The first access log entry is generated one interval after a new connection is received by the TCP Proxy whether or not an upstream connection has been made.
```

Comment (ohadvano) [reply docs/root/intro/arch_overview/observability/access_logging.rst]: done, thanks

**Review by zuercher — APPROVED** [review]
Thanks!

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
